### PR TITLE
Update to write architecture and bug fixes and stability improvements.

### PR DESCRIPTION
**Pull request by derekcollison**
We had a left over artifact from the shared write buffer architecture where the filestore was still controlling the flush loop for the special last message block. This changes it such that each message block is running their own flush loop.

Signed-off-by: Derek Collison <derek@nats.io>

/cc @nats-io/core

### CODE REVIEW

**Review by kozlovic — CHANGES_REQUESTED** [review]
A bit concerned on how the flusher loop is started.

File: server/filestore.go [hunk 1]
```diff
@@ -895,26 +905,20 @@ func (fs *fileStore) removeMsg(seq uint64, secure bool) (bool, error) {
...
+		qch = mb.qch
+		fch = mb.fch
 	}
 	mb.mu.Unlock()
 
 	// Kick outside of lock.
-	if shouldWriteIndex && dch != nil {
+	if shouldWriteIndex {
+		if qch == nil {
+			mb.spinUpFlushLoop()
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/filestore.go:918]: True will set flusher bool in spinup after test.

File: server/filestore.go [hunk 2]
```diff
@@ -959,14 +963,106 @@ func (fs *fileStore) isClosed() bool {
...
 
-// Loop on requests to write out our index file. This is used when calling
-// remove for a message. Updates to the last.seq etc are handled by main
-// flush loop when storing messages.
-func (fs *fileStore) flushWriteIndexLoop(mb *msgBlock, dch, qch chan struct{}) {
+// Will spin up our flush loop.
+func (mb *msgBlock) spinUpFlushLoop() {
+	mb.mu.Lock()
+	// Are we already running?
+	if mb.flusher {
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/filestore.go:970]: Yep will change as described above.

**Comment by derekcollison** [comment]
Posted changes. Please take a look when you have a chance.

File: server/filestore.go [hunk 3]
```diff
@@ -971,6 +971,7 @@ func (mb *msgBlock) spinUpFlushLoop() {
 		mb.mu.Unlock()
 		return
 	}
+	mb.flusher = true
```
Comment (kozlovic) [inline server/filestore.go:974]: May not need the `setInFlusher()` function and its call in flushLoop() anymore?

Comment (derekcollison) [reply server/filestore.go:974]: I looked at that but liked the symmetry there so left it. Can remove it and put comment there if you feel strongly about it.

Comment (kozlovic) [reply server/filestore.go:974]: that's fine.

**Review by kozlovic — APPROVED** [review]

**Comment by kozlovic** [comment]
LGTM

**Pull request merged**